scripts/ligandtocore/QD.py

```diff
@@ -0,0 +1,190 @@
+    """
+    function that handles all core+ligand operations
+    """
+    # Rotate and translate all ligands to their position on the core
+    # Returns a list with sublist [0] containing the rotated ligands (PLAMS Molecules) and
+    # [1] the heteroatoms (PLAMS Atoms) of the rotated ligands to be attached to the core
+    # All core dummy atoms are deleted
+    core = copy.deepcopy(core)
+    ligand_list = [qd_scripts.rotate_ligand(core, ligand, core_index, ligand_index)
+                   for core_index in core_indices]
+
+    ligand_list, ligand_indices = zip(*ligand_list)
+    core.delete_atom(core[-1])
+
+    # Prepare the .pdb filename (string)
+    core_name = core.get_formula()
+
+    ligand_formula = ligand_list[0].get_formula()
+    ligand_heteroatom = ligand[ligand_index + 1].symbol
+    ligand_name = ligand_formula + '_@_' + ligand_heteroatom + str(ligand_index + 1)
+
+    pdb_name = str('core_' + core_name + '___ligand_' + ligand_name)
+
+    # Attach the rotated ligands to the core, returning the resulting strucutre (PLAMS Molecule)
+    qd = qd_scripts.combine_qd(core, ligand_list)
+
+    # indices of all the atoms in the core and the ligand heteroatom anchor
+    qd_indices = [qd.atoms.index(atom) + 1 for atom in ligand_indices]
+    qd_indices += [i + 1 for i, atom in enumerate(core)]
+
+    molkit.writepdb(qd, os.path.join(qd_folder, pdb_name + '.pdb'))
+    print('core + ligands:\t\t\t' + pdb_name + '.pdb')
+
+    return qd, pdb_name, qd_indices
+
+
+def prep_prep(path, dir_name_list, input_cores, input_ligands, smiles_extension, smiles_column,
+              dummy, core_opt, ligand_opt, qd_opt):
+    """
+    function that handles all tasks related to prep_core, prep_ligand and prep_qd
+    """
+    # The start
+    time_start = time.time()
+    print('\n')
+    
+    # managing the result directories
+    core_folder, ligand_folder, qd_folder = [qd_scripts.create_dir(name, path=path)
+                                             for name in dir_name_list]
+ 
+    # Imports the cores and ligands
+    core_list = qd_scripts.read_mol(core_folder, input_cores)
+    ligand_list = qd_scripts.read_mol(ligand_folder, input_ligands,
+                                      smiles_column=smiles_column, smiles_extension=smiles_extension)
+    
+    # Return the indices of the core dummy atoms
+    core_indices = [prep_core(core, core_folder, dummy=dummy, opt=core_opt) for core in core_list]
+    
+    # Open the ligand database and check if the specified ligand(s) is already present
+    database = qd_scripts.read_database(ligand_folder)
+    ligand_list = [prep_ligand(ligand, ligand_folder, database, opt=ligand_opt) for
+                   ligand in ligand_list]
+    
+    # formating of ligand_list
+    ligand_list, ligand_indices, database_entries = zip(*ligand_list)
+    ligand_indices = list(itertools.chain(*ligand_indices))
+    ligand_list = list(itertools.chain(*ligand_list))
```

Comment:
You do not need to transform an iterator to a list if the function that process `ligand_list` accepts *iterators* as arguments. For example, in this case, `ligand_list` is feed into the `enumerate` that can handle iterators.

an example:
```
from itertools import chain
xs = chain(range(3), range(5,8))
rs = [(i, k) for i, k in enumerate(xs)]
#Output: [(0, 0), (1, 1), (2, 2), (3, 5), (4, 6), (5, 7)]
```